server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java

```diff
@@ -51,26 +51,36 @@ private static void put(Map<String, Analyzer> analyzers, String key, Analyzer va
 
     public static MappingLookup fromMapping(Mapping mapping, Analyzer defaultIndex) {
         List<ObjectMapper> newObjectMappers = new ArrayList<>();
-        List<FieldMapper> newFieldMappers = new ArrayList<>();
+        Map<String, FieldMapper> newFieldMappers = new HashMap<>();
         List<FieldAliasMapper> newFieldAliasMappers = new ArrayList<>();
+
+        collect(mapping.root, newObjectMappers, newFieldMappers, newFieldAliasMappers);
+
+        //add runtime fields and replace concrete fields with runtime fields with the same name
+        for (RuntimeFieldMapper runtimeMapper : mapping.root.runtimeMappers()) {
```

Comment:
How does this work when parsing documents? We presumably want to ensure that an incoming doc is parsed using the non-runtime version of the mapper if this exists?

Reply:
good point. I thought that I tested this, but looking at the code I think replacing the field mappers should also affect documents parsing which I think is not what we want.

Reply:
Yeah, I think we need to do this via MappedFieldType, not via FieldMapper, unfortunately.  Which means we need to move `valueFetcher()` - I can make a start on that today.

Reply:
I am not sure that valueFetcher is the only problem, I would be happy if that was the case.

So the idea could be to only replace field mappers that are provided to `FieldTypeLookup`? And go through the usages of `MappingLookup#getMapper` and `MappingLookup#iterator` to see if they may need to access the runtime fields or not?

Reply:
Yes, that sounds reasonable.  Or we could even pass the runtime mappers directly to FieldTypeLookup, like we do currently with aliases, and re-map things in there?

Reply:
yea that would work too, ok I will try to find out what else may be on our way here.